packages/flutter/lib/src/widgets/app.dart

```diff
@@ -610,6 +613,22 @@ class WidgetsApp extends StatefulWidget {
   /// {@endtemplate}
   final RouteInformationProvider? routeInformationProvider;
 
+  /// {@template flutter.widgets.widgetsApp.routeInformationCodec}
+  /// A codec to convert [RouteInformation] into a serializable form.
+  ///
+  /// This codec is used for encoding and decoding route information during
+  /// router's state restoration.
+  ///
+  /// Consider provide a codec if the [RouteInformation.state] may contain
```

Comment:
Currently, we document on `RouteInformation.state` that it must be serializable on its own. If we go down this path, we need to update that.

Reply:
It's currently not clear to me why it is advantageous to allow complex (non-serializable) objects in RouteInformation.state. Isn't the whole point of this property that it will be serialized? Why not do the conversion to serializable state whenever a RouteInformation object is instantiated rather than pushing that complexity into the lower layers by adding a codec property?

Reply:
I am also debating on this. The reason I decide to go with this approach is that not every platform require serialization of state. e.g. desktop apps. For them, the state being serializable is a limitation that impose on them for no reason.

Reply:
Is this limitation actually imposed on Desktop apps? I would have expected that if you never do state restoration (aka you're running on desktop) we actually never attempt to serialize [RouteInformation.state] and everything is sort-of fine even if you use complex objects there? Or do we proactively check somewhere that it can be serialized?